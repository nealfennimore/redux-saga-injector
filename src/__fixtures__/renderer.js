import PropTypes from 'prop-types';
import React from 'react';
import ReactDOMServer from 'react-dom/server';

import { Provider } from 'react-redux';
import createStore from '../__fixtures__/store';
import { preloadQueue } from '../sagas/queue';

const Application = ( {
    store,
    children,
} )=> (
    <Provider store={store}>
        { children }
    </Provider>
);

Application.propTypes = {
    store: PropTypes.shape( {} ).isRequired,
    children: PropTypes.node.isRequired,
};


const renderMarkup = ( App, {store} )=> (
    ReactDOMServer.renderToString(
        <Application store={store}>
            <App store={store} />
        </Application>
    )
);

export default function render( App, store, req, res, options ) {
    return async function() {
        // Server saga listens for any injected sagas to finish
        const preload = store.runSaga( preloadQueue, options );

        // Start initial render to start sagas
        // This is a throw away render
        renderMarkup( App, { store } );

        // Proceed after preload saga finishes
        await preload.done;

        // Get markup with updated store
        const html = renderMarkup( App, { store } );

        res.send( `
            <!DOCTYPE html>
            <html>
                <head>
                </head>
                <body>
                    <div id="app">${html}</div>
                </body>
            </html>
        ` );
    };
}
