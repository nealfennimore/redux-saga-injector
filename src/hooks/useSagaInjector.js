/**
 * External Dependencies
 */
import { useEffect, useState } from 'react';
import {
    useDispatch
} from 'react-redux';

/**
 * Internal Dependencies
 */
import { genUID } from 'src/utils/uid';
import {
    cancelSagas,
    runSagas
} from 'src/actions/sagas';


export default function useSagaInjector( sagas ) {
    if( ! ( sagas && sagas.length ) ) {
        throw new Error( 'No sagas given' );
    }

    const dispatch = useDispatch();
    const [ uid ] = useState( genUID() );

    useEffect(
        ()=>{
            dispatch( runSagas( sagas, uid ) );
            return ()=>{
                dispatch( cancelSagas( uid ) );
            };
        },
        [uid]
    );
}
