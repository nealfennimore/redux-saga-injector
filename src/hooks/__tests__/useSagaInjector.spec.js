import React, {
    Component
} from 'react';
import { useDispatch } from 'react-redux';
import renderer from 'react-test-renderer';
import useSagaInjector from '../useSagaInjector';

jest.mock( 'src/utils/uid', () => ( {
    genUID: () => 'uid'
} ) );

jest.spyOn( React, 'useEffect' ).mockImplementation( f => f() );

const dispatch = jest.fn();
function mockUseDispatch() {
    return dispatch;
}
jest.mock( 'react-redux', ()=>( {
    useDispatch: mockUseDispatch
} ) );

function TestComponent( props ) {
    useSagaInjector( props.sagas );

    return <div>1</div>;
}

describe( 'useSagaInjector', () => {
    let sagas;
    beforeEach( () => {
        sagas = [
            jest.fn(),
            jest.fn(),
            jest.fn()
        ];
        dispatch.mockReset();
    } );

    it( 'should match snapshot', () => {
        const component = renderer.create( <TestComponent sagas={sagas} />
        );
        expect( component.toJSON() ).toMatchSnapshot();
    } );

    describe( 'lifecycle', () => {
        it( 'should inject sagas on mount', () => {
            const component = mount( <TestComponent sagas={sagas} /> );

            expect( component.props().sagas ).toEqual( sagas );
            expect( dispatch.mock.calls ).toMatchSnapshot();
        } );
        it( 'should throw when no sagas', () => {
            expect( ()=> mount( <TestComponent sagas={[]} /> ) ).toThrow( Error );
        } );
        it( 'should unmount properly', () => {
            const component = mount( <TestComponent sagas={sagas} /> );
            component.update();
            component.unmount();
            expect( dispatch.mock.calls ).toMatchSnapshot();
        } );
    } );
} );
