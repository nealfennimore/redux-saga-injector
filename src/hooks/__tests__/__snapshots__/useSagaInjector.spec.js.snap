// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSagaInjector lifecycle should inject sagas on mount 1`] = `
Array [
  Array [
    Object {
      "sagas": Array [
        [MockFunction],
        [MockFunction],
        [MockFunction],
      ],
      "type": "@@redux-saga-injector/RUN_SAGAS",
      "uid": "uid",
    },
  ],
]
`;

exports[`useSagaInjector lifecycle should unmount properly 1`] = `
Array [
  Array [
    Object {
      "sagas": Array [
        [MockFunction],
        [MockFunction],
        [MockFunction],
      ],
      "type": "@@redux-saga-injector/RUN_SAGAS",
      "uid": "uid",
    },
  ],
]
`;

exports[`useSagaInjector should match snapshot 1`] = `
<div>
  1
</div>
`;
